src/pkt_checker.rs

```diff
@@ -0,0 +1,206 @@
+    recv_pkt_map: Arc<RwLock<HashMap<Msn, Arc<Mutex<RecvPktMap>>>>>,
+    read_op_ctx_map: Arc<RwLock<HashMap<Msn, ReadOpCtx>>>,
+}
+
+impl PacketCheckerContext {
+    fn working_thread(ctx: &Self, stop_flag: &AtomicBool) {
+        while !stop_flag.load(Ordering::Relaxed) {
+            if let Err(e) = ctx.check_pkt_map() {
+                error!("PacketChecker stopped: {:?}", e);
+                break;
+            }
+        }
+    }
+    fn check_pkt_map(&self) -> Result<(), Error> {
+        let mut remove_list = LinkedList::new();
+        let iter_maps = {
+            let guard = self
+                .recv_pkt_map
+                .read()
+                .map_err(|_| Error::LockPoisoned("read_op_ctx_map lock"))?;
+            guard
+                .iter()
+                .map(|(k, v)| (*k, Arc::clone(v)))
+                .collect::<Vec<_>>()
+        };
+        for (msn, map) in iter_maps {
+            let (is_complete, is_read_resp, is_out_of_order, dqpn, end_psn) = {
+                let guard = map
+                    .lock()
+                    .map_err(|_| Error::LockPoisoned("recv packet map lock"))?;
+                (
+                    guard.is_complete(),
+                    guard.is_read_resp(),
+                    guard.is_out_of_order(),
+                    guard.dqpn(),
+                    guard.end_psn(),
+                )
+            };
+            // send ack
+            if is_complete {
+                info!("Complete: {:?}", &msn);
+                if !is_read_resp {
+                    // If we are not in read response, we should send ack
+                    let command =
+                        RespCommand::Acknowledge(RespAckCommand::new_ack(dqpn, msn, end_psn));
+                    self.send_queue
+                        .send(command)
+                        .map_err(|_| Error::PipeBroken("packet checker send queue"))?;
+                } else if let Some(ctx) = self
+                    .read_op_ctx_map
+                    .read()
+                    .map_err(|_| Error::LockPoisoned("read_op_ctx_map lock"))?
+                    .get(&msn)
+                {
+                    if let Err(e) = ctx.set_result(()) {
+                        error!("Set result failed {:?}", e);
+                    }
+                } else {
+                    error!("No read op ctx found for {:?}", msn);
+                }
+                remove_list.push_back(msn);
+            } else if is_out_of_order {
+                // TODO: what should we put in NACK packet?
+                let command = RespCommand::Acknowledge(RespAckCommand::new_nack(
+                    dqpn,
+                    Msn::default(),
+                    end_psn,
+                    Psn::default(),
+                ));
+                self.send_queue
+                    .send(command)
+                    .map_err(|_| Error::PipeBroken("packet checker send queue"))?;
+                panic!("send nack command")
```

Comment:
What's the meaning of this panic message? Does it tells the reason of this panic? It just tells where the panic happens.

Reply:
In the previous discussion, we agreed that the first version does not need to implement NACK, and we will panic when NACK needs to be sent. I will add a comment here.